src/nvim/eval.c

```diff
@@ -8430,9 +8429,10 @@ static void filter_map(typval_T *argvars, typval_T *rettv, int map)
   int save_did_emsg;
   int idx = 0;
 
+  list_T *l = NULL;
   if (argvars[0].v_type == VAR_LIST) {
-    if ((l = argvars[0].vval.v_list) == NULL
-        || (!map && tv_check_lock(l->lv_lock, arg_errmsg, TV_TRANSLATE))) {
+    l = argvars[0].vval.v_list;
+    if (l != NULL && !map && tv_check_lock(l->lv_lock, arg_errmsg, TV_TRANSLATE)) {
```

Comment:
I would suggest to leave check as-is. Just move tv_copy() here and into the VAR_DICT case to leave only invalid inputs with zero return, modifications are in-place in any case. This also means that you need not modify anything else at all, only one line transition to two places.